src/PromTracker.mo

```diff
@@ -0,0 +1,291 @@
+    /// ```
+    public func addCounter(prefix : Text, isStable : Bool) : AccumulatorValueRef {
+      let id = Vec.size(values_);
+      let value = AccumulatorValue(prefix, isStable);
+      Vec.add(values_, ?value);
+      {
+        value = value.value;
+        set = value.set;
+        add = value.add;
+        remove = func() = removeValue(id);
+      };
+    };
+
+    /// Add a gauge value for ever changing value, with ability to catch the highest and lowest value during interval, set on tracker instance.
+    /// outputs stats: sum of all pushed values, amount of pushes, lowest value during interval, highest value during interval
+    ///
+    /// Example:
+    /// ```motoko
+    /// let requestDuration = tracker.addGauge("request_duration", true);
+    /// ....
+    /// requestDuration.update(123);
+    /// requestDuration.update(101);
+    /// ```
+    public func addGauge(prefix : Text, isStable : Bool) : GaugeValueRef {
+      let id = Vec.size(values_);
+      let value = GaugeValue(prefix, watermarkResetIntervalSeconds, isStable);
+      Vec.add(values_, ?value);
+      {
+        value = value.lastValue;
+        update = value.update;
+        remove = func() = removeValue(id);
+      };
+    };
+
+    /// Add system metrics, such as cycle balance, memory size, heap size etc.
+    public func addSystemValues() {
+      ignore addPullValue("cycles_balance", func() = Cycles.balance());
+      ignore addPullValue("rts_memory_size", func() = Prim.rts_memory_size());
+      ignore addPullValue("rts_heap_size", func() = Prim.rts_heap_size());
+      ignore addPullValue("rts_total_allocation", func() = Prim.rts_total_allocation());
+      ignore addPullValue("rts_reclaimed", func() = Prim.rts_reclaimed());
+      ignore addPullValue("rts_max_live_size", func() = Prim.rts_max_live_size());
+      ignore addPullValue("rts_max_stack_size", func() = Prim.rts_max_stack_size());
+      ignore addPullValue("rts_callback_table_count", func() = Prim.rts_callback_table_count());
+      ignore addPullValue("rts_callback_table_size", func() = Prim.rts_callback_table_size());
+      ignore addPullValue("rts_mutator_instructions", func() = Prim.rts_mutator_instructions());
+      ignore addPullValue("rts_collector_instructions", func() = Prim.rts_collector_instructions());
+      ignore addPullValue("stablememory_size", func() = Nat64.toNat(StableMemory.size()));
+    };
+
+    /// Dump all current stats to array
+    public func dump() : [(Text, Nat)] {
+      let result : Vec.Vector<(Text, Nat)> = Vec.new();
+      for (v in Vec.vals(values_)) {
+        switch (v) {
+          case (?value) Vec.addFromIter(result, Iter.fromArray(value.dump()));
+          case (null) {};
+        };
+      };
+      Vec.toArray(result);
+    };
+
+    func renderSingle(name : Text, value : Text, timestamp : Text) : Text = name # "{} " # value # " " # timestamp # "\n";
+
+    /// Render all current stats to prometheus format
+    public func renderStats() : Text {
```

Comment:
Let's rename it to `renderExposition` because there is a known term of the text-based "prometheus exposition format".

Reply:
Done